### PR TITLE
Attempt to reduce flakiness of integration tests

### LINKED ISSUE #31722: `TestIntegrations/Disconnection/concurrent_connection_limits_exceeded_node_recording` flakiness
#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/6148508988/job/16682365820

#### Relevant snippet

```
        	            				/__w/teleport/teleport/integration/integration_test.go:2058
        	Error:      	Async error
        	Test:       	TestIntegrations/Disconnection/concurrent_connection_limits_exceeded_node_recording
        	Messages:   	Expected error to contain "administratively prohibited", got: failed connecting to host localhost:37531: failed to receive cluster details response
        	            		failed to dial target host
        	            		direct dialing to nodes not found in inventory is not supported
```

### LINKED ISSUE #33519: `TestIntegrations/DataTransfer` flakiness
#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/6532168318/job/17734847570

#### Relevant snippet

```
    integration_test.go:6216: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:6216
        	            				/__w/teleport/teleport/integration/integration_test.go:123
        	Error:      	Received unexpected error:
        	            	failed connecting to host localhost:40313: failed to receive cluster details response
        	            		failed to dial target host
        	            		direct dialing to nodes not found in inventory is not supported
        	Test:       	TestIntegrations/DataTransfer
```

### LINKED ISSUE #39459: `TestIntegrations/Disconnection/expired_cert_node_recording ` flakiness
@lxea Another one similar to https://github.com/gravitational/teleport/issues/31722 and https://github.com/gravitational/teleport/issues/24358.

https://github.com/gravitational/teleport/actions/runs/8308974520/job/22739814260

```
    integration_test.go:2405: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:2405
        	            				/__w/teleport/teleport/integration/integration_test.go:2284
        	Error:      	Async error
        	Test:       	TestIntegrations/Disconnection/expired_cert_node_recording
        	Messages:   	expected EOF, ExitError, or nil, got failed connecting to host localhost:35667: failed to receive cluster details response
        	            		failed to dial target host
        	            		Direct dialing to nodes not found in the inventory is not supported.
        	            	If you want to connect to a node without installing Teleport on it, consider registering it with
        	            	your cluster with 'teleport join openssh'.
```

### LINKED ISSUE #39515: `TestIntegrations/Disconnection/client_idle_timeout_node_recoding` flakiness
https://github.com/gravitational/teleport/actions/runs/8329492053/job/22791948584

```
=== FAIL: integration TestIntegrations/Disconnection/client_idle_timeout_node_recoding (3.19s)
    integration_test.go:2326: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:2326
        	            				/__w/teleport/teleport/integration/integration_test.go:2213
        	Error:      	Async error
        	Test:       	TestIntegrations/Disconnection/client_idle_timeout_node_recoding
        	Messages:   	expected EOF, ExitError, or nil, got failed connecting to host localhost:41163: failed to receive cluster details response
        	            		failed to dial target host
        	            		direct dialing to nodes not found in inventory is not supported instead
        --- FAIL: TestIntegrations/Disconnection/client_idle_timeout_node_recoding (3.19s)
```

### LINKED ISSUE #46730: `TestIntegrations/PAM/Disabled` flakiness
#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/10928608956/job/30337429955

#### Relevant snippet

```
    integration_test.go:5228: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:5228
        	Error:      	Received unexpected error:
        	            	failed connecting to host localhost:45191: failed to receive cluster details response
        	            		failed to dial target host
        	            		direct dialing to nodes not found in inventory is not supported
        	Test:       	TestIntegrations/PAM/Disabled
```

Hit this in the merge queue on branch/v16

### LINKED ISSUE #47156: `TestIntegrations` flakiness
#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/11168393459/job/31046697056

#### Relevant snippet

```
    integration_test.go:4867: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:4867
        	Error:      	Received unexpected error:
        	            	failed connecting to host localhost:24990: failed to receive cluster details response
        	            		failed to dial target host
        	            		direct dialing to nodes not found in inventory is not supported
        	Test:       	TestIntegrations/ProxyHostKeyCheck/Disabled
```

I'm making this issue to cover multiple subtests of `TestIntegrations` that are all flaky for the same inventory error reason:

- https://github.com/gravitational/teleport/issues/46730
- https://github.com/gravitational/teleport/issues/39515
- https://github.com/gravitational/teleport/issues/39459
- https://github.com/gravitational/teleport/issues/33519
- https://github.com/gravitational/teleport/issues/31722
- https://github.com/gravitational/teleport/issues/24358
- `TestIntegrations/ProxyHostKeyCheck` (this issue)

### PR DESCRIPTION
**Pull request by rosstimothy**
Closes #47156.
Closes https://github.com/gravitational/teleport/issues/48823.
Closes https://github.com/gravitational/teleport/issues/46730.
Closes https://github.com/gravitational/teleport/issues/39459.
Closes https://github.com/gravitational/teleport/issues/33519.
Closes https://github.com/gravitational/teleport/issues/31722.
Closes https://github.com/gravitational/teleport/issues/24358.
Closes https://github.com/gravitational/teleport/issues/39515.

All of the tests suffering from issues dialing hosts, and failing with a `failed to dial target host` error were incorrectly waiting for nodes to become visible before establishing connections. The main culprit for most of the failures was `waitForNodesToRegister`, though a few tests had a very similar hand rolled variant, which incorrectly returned when the nodes appeared in Auth. However, since the Proxy is the one performing dialing, they should have waited for the nodes to appear in the Proxy.

To resolve, `waitForNodesToRegister` and all hand rolled equivalents have been removed in favor of `helpers.WaitForNodeCount` which correctly uses the `CachingAccessPoint` of the RemoteSite instead of `GetClient`. Additionally, `helpers.WaitForNodeCount` was updated to validate that the node watcher used for routing in the Proxy also contained the expected number of nodes.

### CODE REVIEW

File: integration/helpers/trustedclusters.go [hunk 1]
```diff
@@ -124,18 +124,32 @@ func WaitForNodeCount(ctx context.Context, t *TeleInstance, clusterName string,
...
+		if len(nodes) != count {
+			return trace.BadParameter("cache contained %v nodes, but wanted to find %v nodes", len(nodes), count)
 		}
-		return trace.BadParameter("found %v nodes, but wanted to find %v nodes", len(nodes), count)
+
+		// Validate that the site watcher contains the expected count.
+		watcher, err := remoteSite.NodeWatcher()
+		if err != nil {
+			return trace.Wrap(err)
+		}
```
Comment (hugoShaka) [inline integration/helpers/trustedclusters.go:146]: So without this it was possible to for a node to land into the access point's cache but not being processed yet by the node watcher?

File: integration/helpers/trustedclusters.go
Comment (hugoShaka) [inline integration/helpers/trustedclusters.go:1]: this PR consolidates every node wait into `WaitForNodeCount`. This function doesn't seem trustedcluster-specific, should we move it to another file?

**Review by hugoShaka — APPROVED** [review]

File: integration/helpers/helpers.go [hunk 2]
```diff
@@ -550,3 +550,49 @@ func UpsertAuthPrefAndWaitForCache(
 		assert.Empty(t, cmp.Diff(&pref, &p))
 	}, 5*time.Second, 100*time.Millisecond)
 }
+
+// WaitForNodeCount waits for a certain number of nodes to be visible
+// to the Proxy in clusterName.
+func WaitForNodeCount(ctx context.Context, t *TeleInstance, clusterName string, count int) error {
```
Comment (zmb3) [inline integration/helpers/helpers.go:556]: Would this be more discoverable if it were a method on `TeleInstance`?

Can we get the cluster name from `t.Process.Config.Auth.ClusterName`, or do we sometimes have to use other cluster names?

**Review by zmb3 — APPROVED** [review]
Look at that list of closed issues!

**Pull request merged**